### ISSUE TITLE
Should I use body_filter_by_lua_block to perform filter on the entire response body?

### ISSUE THREAD

**Issue by hariszukanovic** [openresty/lua-nginx-module#1813]
The question is general in nature, but I work with version openresty/1.17.8.2

I need to filter (search and replace some strings) in the response body possibly arriving from upstream...
From docs, I understand that the response body received in ngx.arg[1] can arrive split into multiple parts i.e. body_filter_by_lua_block is called multiple times with those parts.

I found many simple usage examples that suggest it exactly done like in the example below... so I wonder if I am missing something and they actually can work correctly after all?

Namely, it would seem that filtering and using the string.gsub cannot possibly work correctly, as long as there is the possibility that an instance of 'will-be-replaced-text' is actually split somewhere in the middle between the calls. It would not be matched when ngx.arg[1] is given as subject string to string.gsub

Example:
This example would produce response exactly like the original without substituting anything.

Response body:
This is a **will-be-replaced-text**, and it is split in an unlucky place for me...

Immagine it arrive split in 2 parts like this:
1. This is a **will-be-repla**
2.'**ced-text**, and it is split in an unlucky place for me...

body_filter_by_lua_block {                
                ngx.arg[1] = string.gsub(ngx.arg[1], 'will-be-replaced-text', 'replace-with-some-other-text')
            }

I know that in my case the response body is small enough not to cause memory problems for me...
Is there a way to ensure that I receive the whole response body in 1 piece?
Or is there any other way to correctly search and replace something the response body?

**Comment by hariszukanovic** [comment]
Hi again,
I was hoping for any kind of comment...
But, it does look like the body_filter_by_lua* is actually useful only if you wish to filter on character basis?

**Comment by rhuddleston** [comment]
I'm running into this same issue, any suggested workaround?

**Comment by deba12** [comment]
Yes buffer whole response in ngx.ctx.body and then run gsub on it.

pseudo code
```
ngx.ctx.body = ngx.ctx.body .. ngx.arg[1]

if ngx.arg[2] == true then
string.gsub(ngx.ctx.body, z...)
end
```

**Comment by rhuddleston** [comment]
Fully tested version for others interested:

`ngx.ctx.buffered = (ngx.ctx.buffered or "") .. ngx.arg[1]
if ngx.arg[2] then
    ngx.arg[1] = ngx.ctx.buffered:gsub('CDN', 'replace_string')
else
    ngx.arg[1] = ''
end`

**Comment by Neolo** [comment]
> Fully tested version for others interested:
> 
> ```
> ngx.ctx.buffered = (ngx.ctx.buffered or "") .. ngx.arg[1]
> if ngx.arg[2] then
>     ngx.arg[1] = ngx.ctx.buffered:gsub('CDN', 'replace_string')
> else
>     ngx.arg[1] = ''
> end
> ```

It cuts off 6 last characters of the body on output

**Comment by zhuizhuhaomeng** [comment]
any example to reproduce this?

**Comment by Neolo** [comment]
> any example to reproduce this?

Adding header_filter_by_lua_block { ngx.header.content_length = nil }
fixed the issue.

Reproduced in cli curl output.

**Comment by doujiang24** [comment]
> Adding header_filter_by_lua_block { ngx.header.content_length = nil }
> fixed the issue.

Yeah, this is expected. We should force the response body to chunked format.

**Comment by Neolo** [comment]
> > Adding header_filter_by_lua_block { ngx.header.content_length = nil }
> > fixed the issue.
> 
> Yeah, this is expected. We should force the response body to chunked format.

But that is not good. Moreover, I had to force header Transfer-Encoding: chunked as it's not set automatically in this case.
How to set a correct content-length after modifier instead?

**Comment by doujiang24** [comment]
> How to set a correct content-length after modifier instead?

I don't think there is a good way to approach it unless we break the nginx stream pipeline.
In the current implementation, the response header has already been sent out before handing the response body in the body_filter_by_lua.

**Comment by doujiang24** [comment]
> Moreover, I had to force header Transfer-Encoding: chunked as it's not set automatically in this case. And disable keep-alive.

Also, I have tested only `set content-length to nil` is works, with such a nginx.conf.
https://gist.github.com/doujiang24/547d4ec05e6c980459142abb1f4248cb

and here is the test result:

```
# original output:
$ curl 'http://127.0.0.1:8099/hello' -i
HTTP/1.1 200 OK
Server: openresty/1.19.9.1
Date: Fri, 24 Sep 2021 05:17:33 GMT
Content-Type: application/octet-stream
Connection: keep-alive
Content-Length: 6

hello

# modifier
$ curl 'http://127.0.0.1:8099/modifier' -i
HTTP/1.1 200 OK
Server: openresty/1.19.9.1
Date: Fri, 24 Sep 2021 05:17:27 GMT
Content-Type: application/octet-stream
Transfer-Encoding: chunked
Connection: keep-alive

he====o
```

**Comment by Neolo** [comment]
Client doesn't read chunked.

How about ngx.location.capture ? How to make it work with body replacement?

**Comment by Neolo** [comment]
Here is a dirty solution to avoid chunked. Use ngx.location.capture on top of another location, where we perform body altering, and then set correct content size. Passing to that location full URI and ARGS.

location / {
        content_by_lua_block {
        local ar = ngx.req.get_uri_args()
        local res = ngx.location.capture('/luapipe'..ngx.var.uri..'?'..ngx.encode_args(ar), { method = ngx.HTTP_GET, body = ngx.var.request_body, copy__all_vars = true })
        if res then
              ngx.header.content_length = #res.body
              ngx.print(res.body)
        end
     }
}

location /luapipe/ {
body_filter_by_lua_block
               {
                ngx.ctx.buffered = (ngx.ctx.buffered or "") .. ngx.arg[1]
                if ngx.arg[2] then
                   ngx.arg[1] = ngx.ctx.buffered:gsub('SEARCH_BODY', 'REPLACE_WITH')
                        else
                            ngx.arg[1] = ''
                end
             }
        proxy_pass http://IP:PORT/;
     }
}
